### PR TITLE
[WIP] array internalization rewrite pass

### PR DESCRIPTION
**Pull request by sklam**
Implement rewrite for CUDA backend to rewrite array allocations into per-thread static local memory of the function.  This allows the use of `np.empty` to be used inside the CUDA backend.  

Limitations:
- only rewrite the array allocations in the entry block
- the shape must be compile-time constant.

### CODE REVIEW

File: numba/array_intern.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            # Install a lowering hook if we are using this rewrite.
+            special_ops = self.pipeline.targetctx.special_ops
+            if 'interned_alloc' not in special_ops:
+                special_ops['interned_alloc'] = _lower_interned_alloc
+
+    def match(self, label, block, typemap, calltypes):
+        if not self._enabled:
+            return False
+        # Match only the first block
+        if label == 0 and block.body and len(calltypes) > 1:
```
Comment (pitrou) [inline numba/array_intern.py]: I'm wondering if we can find something better than `label == 0` (which sounds a bit obscure) to test for this.

File: numba/array_intern.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+from __future__ import print_function, division, absolute_import
+import numpy as np
+from numba import cgutils, ir, types, rewrites
+from numba.typing.templates import FunctionTemplate
+
+
+@rewrites.register_rewrite
+class InternArrayAlloc(rewrites.Rewrite):
```
Comment (pitrou) [inline numba/array_intern.py]: A docstring perhaps?

File: numba/array_intern.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        repls = [newstmts.get(i, stmt)
+                 for i, stmt in enumerate(self._block.body)]
+        newblk = ir.Block(scope=self._block.scope, loc=self._block.loc)
+        newblk.body.extend(repls)
+        return newblk
+
+
+def _lower_interned_alloc(lower, expr):
+    if expr.ty.layout != 'C':
+        raise NotImplementedError("only support C layout")
```
Comment (pitrou) [inline numba/array_intern.py]: Can't this be done at typing time? Or at least in the rewrite phase above?

File: numba/cuda/npydecl.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+from numba.typing.npydecl import NdEmpty
+
+registry = Registry()
+intrinsic = registry.register
+intrinsic_attr = registry.register_attr
+intrinsic_global = registry.register_global
+
+
+@intrinsic_attr
+class NumpyModuleAttribute(AttributeTemplate):
```
Comment (pitrou) [inline numba/cuda/npydecl.py]: I'm not sure this one is needed, is it? register_global() should be sufficient.

**Comment by sklam** [comment]
The usefulness of this feature is questionable due to the unsafe nature of the static array

**Pull request closed without merging**